### DISCUSSION TITLE
Golang实现一个事务型内存数据库

### DISCUSSION THREAD

**Discussion by giscus** [qingwave/qingwave.github.io#81] [Announcements]
# Golang实现一个事务型内存数据库

内存数据库经我们经常用到，例如Redis，那么如何从零实现一个内存数据库呢，本文旨在介绍如何使用Golang编写一个KV内存数据库MossDB。 特性MossDB是一个纯Golang编写、可嵌入的、键值

https://qingwave.github.io/golang-in-memory-database/

**Comment by acechef** [comment]
厉害厉害

**Reply by acechef** [reply]
已start

**Reply by qingwave** [reply]
哈哈，多谢支持

**Comment by rfyiamcool** [comment]
😁 事务的设计有点小问题，比如，在事务里发起了更新操作，还未提交，但其他并发事务是可以脏读的。并发事务下也会有读已提交的现象，当然也就是实现 RC 那没问题。你其实需要解决多版本 MVCC，一致性视图的问题，不然在一个事务中一会读到 a=1，一会读到 a=2 ，重复读没法保证。

可以参考 golang badger 和 boltdb 的事务实现，boltdb 的实现相对简单点，写写阻塞，badger 的事务实现要复杂一些。

wal 需要模拟 redis aof rewrite ，其实可以设计成 bitcask 模式，这样无需像 redis aof rewrite 那样，每次都把内存数据全量写到文件里。

watch 的实现挺好的，但你这 workid 的实现是不是有点复杂拉，直接一个时间戳或者 atomic int 完事。

**Reply by qingwave** [reply]
感谢建议，难得有这么中肯深刻的评论了:smile:

事务设计上为了实现简单，使用的是全局锁，boltdb事务实现了读写两个接口，这里为了保证API简洁，只提供了统一的事务API。如果要支持并发事务，确实得考虑MVCC。

大概看了bitcask的实现，确实挺不错的，后续可以借鉴下。

再次感谢:)

**Reply by rfyiamcool** [reply]
其实不需要回滚的，复杂度加大，直接把事务改的简单点。

事务中的写都写到 txn buffer 里，把事务中读到的数据也放到 txn cache 里，只在 commit 提交时进行 db 加锁做写提交，而事务中的读优先在 txn cache -> buffer 中去读，这样也可以做到简单的一致性读视图。也可以进一步实现乐观事务，这里需要冲突检测，把进行中的事务涉及到的 write keys 都放在一个全局容器里，只要事务中有互相的冲突就直接 abort 。 

😁 先前写过 bitcask 的设计实现系列的文章，有兴趣可以看看。

https://github.com/rfyiamcool/notes#golang-bitcask-rosedb

**Reply by qingwave** [reply]
第一版设计的时候，确实引入了buffer，但后面为了支持前缀查询，buffer的实现过于复杂就没有深入研究。
文章mark了，后面学习学习:smile:
